### PR TITLE
fix: new projects not appearing in create agreement project dropdown

### PR DESCRIPTION
**Pull request by fpigeonjr**
## What changed

Fixes a bug where newly created research projects would not appear in the "Select a Project" dropdown on Step 1 of the Create Agreement wizard, even though they were visible on the Projects list page.

Two separate RTK Query issues were found and fixed — both are frontend-only changes, no backend modifications:

**1. Stale cache on re-mount (`StepSelectProject.jsx`)**
`useGetResearchProjectsQuery` was called without `refetchOnMountOrArgChange: true`. RTK Query's default behaviour caches the response and serves it on re-mount, meaning a user who navigated away to create a project and came back would see the old list. Added `{ refetchOnMountOrArgChange: true }` — the same pattern used in `ChangeRequestsList`, `CanCard`, `PortfolioSpending`, etc.

**2. Backend default `limit=10` silently truncating results (`opsAPI.js`)**
`getResearchProjects` called `/projects/?project_type=RESEARCH` with no `limit` parameter. The backend `PaginationListSchema` defaults to `limit=10`, so only ever the first 10 research projects were returned — explaining the staging discrepancy of 10 in the dropdown vs 15 in the Projects list (5 are non-RESEARCH type, the rest were truncated).

Rather than raising the backend's `max=50` cap (which would deviate from every other endpoint's established pattern), `getResearchProjects` was converted from `query` + `transformResponse` to a `queryFn` pagination loop that fetches all research projects in 50-per-batch requests. This stays within the existing backend contract and scales to any future project count.

## Issue

_No linked issue — reported via Slack._

## How to test

1. Log in and navigate to **Create Agreement**
2. Note which projects appear in the Step 1 "Select a Project" dropdown
3. Open a new tab → **Projects** → create a new **Research** project
4. Navigate back to **Create Agreement** → Step 1
5. ✅ The newly created project should now appear in the dropdown

## A11y impact

- [x] No accessibility-impacting changes in this PR

## Screenshots

SCR-20260429-nldl

## Definition of Done Checklist
- [x] OESA: Code refactored for clarity
- [x] OESA: Dependency rules followed
- [x] Automated unit tests updated and passed
- [ ] Automated integration tests updated and passed
- [ ] Automated quality tests updated and passed
- [ ] Automated load tests updated and passed
- [ ] Automated a11y tests updated and passed
- [ ] Automated security tests updated and passed
- [x] 90%+ Code coverage achieved
- [ ] Form validations updated

## Links

_N/A_

### CODE REVIEW

File: frontend/src/api/opsAPI.js [hunk 1]
```diff
@@ -440,14 +440,37 @@ export const opsApi = createApi({
...
-                    return response.data;
+            // The backend caps every list endpoint at limit=50 (PaginationListSchema).
+            // Rather than deviating from that constraint we page through all results
+            // here so the Create Agreement project dropdown always shows every
+            // research project regardless of how many exist.
+            async queryFn(_arg, _queryApi, _extraOptions, fetchWithBQ) {
+                const BATCH_SIZE = 50;
+                const allProjects = [];
+                let offset = 0;
+                let total = Infinity; // set on first wrapped response
```
Comment (fpigeonjr) [inline frontend/src/api/opsAPI.js:451]: to Infinity and beyond 😄

**Review by johndeange — APPROVED** [review]
## Overview

Fixes two frontend bugs preventing newly-created research projects from showing in the Create Agreement Step 1 dropdown:

1. `StepSelectProject.jsx` — adds `refetchOnMountOrArgChange: true` so re-mount triggers a refetch.
2. `opsAPI.js` — replaces the single-shot `getResearchProjects` with a `queryFn` pagination loop that walks the backend in 50-item batches (respecting `PaginationListSchema`'s `max=50`).

Verified the backend cap at `backend/ops_api/ops/schemas/pagination.py:11` — the 10/15 staging mismatch is accurately diagnosed.

## Strengths

- Root-cause framing is correct; `load_default=[10]` in `PaginationListSchema` really is what truncated results.
- `refetchOnMountOrArgChange` matches the existing project convention (`ChangeRequestsList`, `CanCard`, `PortfolioSpending`).
- Keeps the 50-item cap intact rather than widening the public contract for every list endpoint — the right call.
- Legacy array-format fallback retained, so the wrapped/unwrapped response transition isn't broken.
- Solid new test file covering loading / error / success / empty / cancel flows, and asserting the hook is called with the cache-bust option.

## Suggestions (non-blocking)

**1. Potential infinite loop on malformed `count` (`opsAPI.js:443-472`)**

```js
while (allProjects.length < total) {
    ...
    total = response.count ?? page.length;
    allProjects.push(...page);
    offset += BATCH_SIZE;
}
```

If the backend ever returns `count > allProjects.length` but an empty `page`, the loop spins forever. Belt-and-braces:

```js
if (page.length === 0) break;
```

**2. Tag invalidation should already handle this**

`addResearchProjects` already has `invalidatesTags: ["ResearchProjects"]` (`opsAPI.js:570`), and `getResearchProjects` provides that tag — re-mount should already repopulate. If tag invalidation isn't firing after project creation, that's a latent bug worth documenting. `refetchOnMountOrArgChange: true` is reasonable belt-and-braces, but worth confirming the underlying invalidation is healthy.

**3. No unit test for the pagination `queryFn` itself**

The new test mocks `useGetResearchProjectsQuery`, so the multi-page loop, legacy-array branch, and error short-circuit in `queryFn` are untested. An MSW-backed test or direct `queryFn` invocation test would harden the change most likely to regress (off-by-one on `offset`, empty-page edge case, missing `count`).

**4. Minor: misleading comment (`opsAPI.js:441-444)**

"The backend caps every list endpoint at limit=50" — the schema caps at 50, but not every endpoint uses `PaginationListSchema`. Tighter: *"`/projects/` is capped at limit=50 by `PaginationListSchema`…"*

**5. Test file nit**

`beforeEach` is used without being imported in the `from "vitest"` destructure. Works if Vitest globals are enabled, but inconsistent with the other symbols that *are* imported.

## Security / Performance

- No new auth, serialisation, or input-validation surface. Nothing flagged.
- For current scale (~15 research projects), one extra round trip max. If counts grow into the hundreds, parallelising subsequent batches after the first `count` arrives is a future optimisation — not a now-concern.

## Verdict

Approving. The fix is correct and well-scoped. Recommend adding the empty-page guard (#1) and at least one `queryFn`-level test (#3) before merge — both are small and would make the pagination loop much safer.

File: frontend/src/api/opsAPI.js (hunk 1 above)
Comment (fpigeonjr) [reply frontend/src/api/opsAPI.js:451]: 😄 To Infinity and beyond! It's the cleanest sentinel for 'we don't know the total yet' — gets replaced on the first wrapped response.

[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/api/opsAPI.js]: Added four tests to `opsAPI.test.js` covering: single-batch wrapped response, multi-batch pagination (51 projects across 2 requests), legacy array short-circuit, and API error propagation.

File: frontend/src/pages/agreements/StepSelectProject.jsx [hunk 2]
```diff
@@ -46,7 +46,11 @@ export const StepSelectProject = ({
     const [modalProps, setModalProps] = React.useState({});
     const { setAlert } = useAlert();
     /** @type {{data?: import("../../types/ProjectTypes").Project[] | undefined, error?: Object, isLoading: boolean}} */
-    const { data: projects, error: errorProjects, isLoading: isLoadingProjects } = useGetResearchProjectsQuery({});
+    const {
+        data: projects,
+        error: errorProjects,
+        isLoading: isLoadingProjects
+    } = useGetResearchProjectsQuery({}, { refetchOnMountOrArgChange: true });
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/pages/agreements/StepSelectProject.jsx]: Fixed — changed to `undefined` to share the cache key with `lookup.hooks.js` which calls `useGetResearchProjectsQuery()` with no arg.

File: frontend/src/pages/agreements/StepSelectProject.test.jsx [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        it("calls useGetResearchProjectsQuery with refetchOnMountOrArgChange: true", () => {
+            useGetResearchProjectsQuery.mockReturnValue({
+                data: MOCK_PROJECTS,
+                error: undefined,
+                isLoading: false
+            });
+
+            renderComponent();
+
+            expect(useGetResearchProjectsQuery).toHaveBeenCalledWith({}, { refetchOnMountOrArgChange: true });
```
[the remark this answers is not in the document]
Comment (fpigeonjr) [reply frontend/src/pages/agreements/StepSelectProject.test.jsx]: Updated the assertion to `toHaveBeenCalledWith(undefined, { refetchOnMountOrArgChange: true })` to match the fix.

**Review by fpigeonjr — COMMENTED** [review]
Thanks @johndeange! Addressed in the latest commit:

**#1 — Infinite loop guard:** Added `if (page.length === 0) break;` after the page is unpacked, plus a matching test that verifies a malformed `count: 99` with an empty `data: []` returns cleanly without looping.

**#2 — Tag invalidation:** Agreed — `invalidatesTags: ["ResearchProjects"]` should handle post-creation freshness. `refetchOnMountOrArgChange: true` is intentional belt-and-braces to cover the cross-tab case (user creates a project in a different browser tab/session where the mutation never fires in this Redux store). Both layers working together is the safest approach.

**#3 — queryFn tests:** Already added in the prior commit — 4 MSW-backed tests in `opsAPI.test.js` covering single-batch, multi-batch, legacy array, and error cases. Added a 5th for the empty-page guard (#1).

**#4 — Comment wording:** Tightened to ```/projects/` is capped at limit=50 by `PaginationListSchema````.

**#5 — `beforeEach` import:** Added to the vitest destructure in `StepSelectProject.test.jsx`.

**Comment by fpigeonjr** [comment]
Thanks @johndeange! Addressed in the latest commit:

**#1 — Infinite loop guard:** Added `if (page.length === 0) break;` after the page is unpacked, plus a matching test that verifies a malformed `count: 99` with an empty `data: []` returns cleanly without looping.

**#2 — Tag invalidation:** Agreed — `invalidatesTags: ["ResearchProjects"]` should handle post-creation freshness. `refetchOnMountOrArgChange: true` is intentional belt-and-braces to cover the cross-tab case (user creates a project in a different browser tab/session where the mutation never fires in this Redux store). Both layers working together is the safest approach.

**#3 — queryFn tests:** Already added in the prior commit — 4 MSW-backed tests in `opsAPI.test.js` covering single-batch, multi-batch, legacy array, and error cases. Added a 5th for the empty-page guard (#1).

**#4 — Comment wording:** Tightened to `/projects/` is capped at limit=50 by `PaginationListSchema`.

**#5 — `beforeEach` import:** Added to the vitest destructure in `StepSelectProject.test.jsx`.

**Review by weimiao67 — APPROVED** [review]
